staging/src/k8s.io/apiextensions-apiserver/pkg/apis/apiextensions/validation/validation.go

```diff
@@ -162,91 +162,161 @@ func ValidateCustomResourceDefinitionNames(names *apiextensions.CustomResourceDe
 	}
 
-	if JSONSchema.Ref != "" {
-		openapiRef, err := spec.NewRef(JSONSchema.Ref)
+	if schema.Ref != nil {
+		openapiRef, err := spec.NewRef(*schema.Ref)
 		if err != nil {
-			allErrs = append(allErrs, field.Invalid(fldPath.Child("Ref"), JSONSchema.Ref, err.Error()))
+			allErrs = append(allErrs, field.Invalid(fldPath.Child("ref"), *schema.Ref, err.Error()))
 		}
+
 		if !openapiRef.IsValidURI() {
-			allErrs = append(allErrs, field.Invalid(fldPath.Child("Ref"), JSONSchema.Ref, "Ref does not point to a valid URI."))
+			allErrs = append(allErrs, field.Invalid(fldPath.Child("ref"), *schema.Ref, "ref does not point to a valid URI"))
 		}
 	}
 
-	if JSONSchema.AdditionalItems != nil {
-		allErrs = append(allErrs, ValidateCustomResourceDefinitionSchema(JSONSchema.AdditionalItems.Schema, fldPath.Child("AdditionalItems"))...)
+	if schema.AdditionalItems != nil {
+		allErrs = append(allErrs, ValidateCustomResourceDefinitionOpenAPISchema(schema.AdditionalItems.Schema, fldPath.Child("additionalItems"), ssv)...)
 	}
 
-	allErrs = append(allErrs, ValidateCustomResourceDefinitionSchema(JSONSchema.Not, fldPath.Child("Not"))...)
+	allErrs = append(allErrs, ValidateCustomResourceDefinitionOpenAPISchema(schema.Not, fldPath.Child("not"), ssv)...)
 
-	if len(JSONSchema.AllOf) != 0 {
-		for _, jsonSchema := range JSONSchema.AllOf {
-			allErrs = append(allErrs, ValidateCustomResourceDefinitionSchema(&jsonSchema, fldPath.Child("AllOf"))...)
+	if len(schema.AllOf) != 0 {
+		for _, jsonSchema := range schema.AllOf {
+			allErrs = append(allErrs, ValidateCustomResourceDefinitionOpenAPISchema(&jsonSchema, fldPath.Child("allOf"), ssv)...)
 		}
 	}
 
-	if len(JSONSchema.OneOf) != 0 {
-		for _, jsonSchema := range JSONSchema.OneOf {
-			allErrs = append(allErrs, ValidateCustomResourceDefinitionSchema(&jsonSchema, fldPath.Child("OneOf"))...)
+	if len(schema.OneOf) != 0 {
+		for _, jsonSchema := range schema.OneOf {
+			allErrs = append(allErrs, ValidateCustomResourceDefinitionOpenAPISchema(&jsonSchema, fldPath.Child("oneOf"), ssv)...)
 		}
 	}
 
-	if len(JSONSchema.AnyOf) != 0 {
-		for _, jsonSchema := range JSONSchema.AnyOf {
-			allErrs = append(allErrs, ValidateCustomResourceDefinitionSchema(&jsonSchema, fldPath.Child("AnyOf"))...)
+	if len(schema.AnyOf) != 0 {
+		for _, jsonSchema := range schema.AnyOf {
+			allErrs = append(allErrs, ValidateCustomResourceDefinitionOpenAPISchema(&jsonSchema, fldPath.Child("anyOf"), ssv)...)
 		}
 	}
 
-	if len(JSONSchema.Properties) != 0 {
-		for property, jsonSchema := range JSONSchema.Properties {
-			allErrs = append(allErrs, ValidateCustomResourceDefinitionSchema(&jsonSchema, fldPath.Child("Properties."+property))...)
+	if len(schema.Properties) != 0 {
+		for property, jsonSchema := range schema.Properties {
+			allErrs = append(allErrs, ValidateCustomResourceDefinitionOpenAPISchema(&jsonSchema, fldPath.Child("properties["+property+"]"), ssv)...)
```

Comment:
we have this https://github.com/kubernetes/apimachinery/blob/master/pkg/util/validation/field/path.go#L64 :)